ipa-core/src/query/processor.rs

```diff
@@ -169,44 +175,51 @@ impl Processor {
             config: req,
             roles: roles.clone(),
         };
-
-        // Inform other parties about new query. If any of them rejects it, this join will fail
+        // Inform other helpers about new query. If any of them rejects it, this join will fail
         try_join(
             transport.send(left, prepare_request.clone(), stream::empty()),
             transport.send(right, prepare_request.clone(), stream::empty()),
         )
         .await
         .map_err(NewQueryError::MpcTransport)?;
 
+        shard_transport.broadcast(prepare_request.clone()).await?;
+
         handle.set_state(QueryState::AwaitingInputs(query_id, req, roles))?;
 
         guard.restore();
         Ok(prepare_request)
     }
 
     /// On prepare, each follower:
-    /// * ensures that it is not the leader on this query
+    /// * ensures that it is not the leader helper on this query
     /// * query is not registered yet
-    /// * creates gateway and network
     /// * registers query
     ///
     /// ## Errors
     /// if query is already running or this helper cannot be a follower in it
-    pub fn prepare(
+    pub async fn prepare_helper(
         &self,
-        transport: &MpcTransportImpl,
+        mpc_transport: MpcTransportImpl,
+        shard_transport: ShardTransportImpl,
         req: PrepareQuery,
     ) -> Result<(), PrepareQueryError> {
-        let my_role = req.roles.role(transport.identity());
+        let my_role = req.roles.role(mpc_transport.identity());
+        let shard_index = shard_transport.identity();
 
         if my_role == Role::H1 {
             return Err(PrepareQueryError::WrongTarget);
         }
+        if shard_index != ShardIndex::FIRST {
+            return Err(PrepareQueryError::WrongTarget);
+        }
```

Comment:
```suggestion
        if my_role == Role::H1 || shard_index != ShardIndex::FIRST {
            return Err(PrepareQueryError::WrongTarget);
        }        
```

although it may be better to use a different error message when someone submits prepare request to wrong shard

Reply:
I'll add a separate type. I rather contribute with the error granularity.